<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title></title>
  <meta name="description" content="">
  <meta name="generator" content="bookdown 0.5.10 and GitBook 2.6.7">

  <meta property="og:title" content="" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Little</a></li>
<li class="chapter" data-level="2" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html"><i class="fa fa-check"></i><b>2</b> DNA Sequence Statistics (1)</a><ul>
<li class="chapter" data-level="2.1" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>2.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="2.2" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>2.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="2.3" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#fasta-format"><i class="fa fa-check"></i><b>2.3</b> FASTA format</a></li>
<li class="chapter" data-level="2.4" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>2.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="2.5" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#retrieving-genome-sequence-data-via-the-ncbi-website"><i class="fa fa-check"></i><b>2.5</b> Retrieving genome sequence data via the NCBI website</a></li>
<li class="chapter" data-level="2.6" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#retrieving-genome-sequence-data-using-seqinr"><i class="fa fa-check"></i><b>2.6</b> Retrieving genome sequence data using SeqinR</a></li>
<li class="chapter" data-level="2.7" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#writing-sequence-data-out-as-a-fasta-file"><i class="fa fa-check"></i><b>2.7</b> Writing sequence data out as a FASTA file</a></li>
<li class="chapter" data-level="2.8" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#reading-sequence-data-into-r"><i class="fa fa-check"></i><b>2.8</b> Reading sequence data into R</a></li>
<li class="chapter" data-level="2.9" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>2.9</b> Length of a DNA sequence</a></li>
<li class="chapter" data-level="2.10" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>2.10</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="2.11" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>2.11</b> GC Content of DNA</a></li>
<li class="chapter" data-level="2.12" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#dna-words"><i class="fa fa-check"></i><b>2.12</b> DNA words</a></li>
<li class="chapter" data-level="2.13" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#summary"><i class="fa fa-check"></i><b>2.13</b> Summary</a></li>
<li class="chapter" data-level="2.14" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#links-and-further-reading"><i class="fa fa-check"></i><b>2.14</b> Links and Further Reading</a></li>
<li class="chapter" data-level="2.15" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#acknowledgements"><i class="fa fa-check"></i><b>2.15</b> Acknowledgements</a></li>
<li class="chapter" data-level="2.16" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#contact"><i class="fa fa-check"></i><b>2.16</b> Contact</a></li>
<li class="chapter" data-level="2.17" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#license"><i class="fa fa-check"></i><b>2.17</b> License</a></li>
<li class="chapter" data-level="2.18" data-path="dna-sequence-statistics-1.html"><a href="dna-sequence-statistics-1.html#exercises"><i class="fa fa-check"></i><b>2.18</b> Exercises</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dna-sequence-statistics-1" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> DNA Sequence Statistics (1)</h1>
<div id="using-r-for-bioinformatics" class="section level2">
<h2><span class="header-section-number">2.1</span> Using R for Bioinformatics</h2>
<p>This booklet tells you how to use the R software to carry out some simple analyses that are common in bioinformatics. In particular, the focus is on computational analysis of biological sequence data such as genome sequences and protein sequences.</p>
<p>This booklet assumes that the reader has some basic knowledge of biology, but not necessarily of bioinformatics. The focus of the booklet is to explain simple bioinformatics analysis, and to explain how to carry out these analyses using R.</p>
<p>To use R, you first need to start the R program on your computer. You should have already installed R on your computer (if not, for instructions on how to install R, see <code>How to install R &lt;./installr.html&gt;</code>_).</p>
</div>
<div id="r-packages-for-bioinformatics-bioconductor-and-seqinr" class="section level2">
<h2><span class="header-section-number">2.2</span> R packages for bioinformatics: Bioconductor and SeqinR</h2>
<p>Many authors have written R packages for performing a wide variety of analyses. These do not come with the standard R installation, but must be installed and loaded as “add-ons”.</p>
<p>Bioinformaticians have written several specialised <em>packages</em> for R. In this practical, you will learn to use the SeqinR package to retrieve sequences from a DNA sequence database, and to carry out simple analyses of DNA sequences.</p>
<p>Some well known bioinformatics packages for R are the Bioconductor set of R packages<br />
(<code>www.bioconductor.org &lt;http://www.bioconductor.org/&gt;</code><em>), which contains several packages with many R functions for analysing biological data sets such as microarray data; and the SeqinR package (<code>pbil.univ-lyon1.fr/software/seqinr/home.php?lang=eng &lt;http://pbil.univ-lyon1.fr/software/seqinr/home.php?lang=eng&gt;</code></em>), which contains R functions for obtaining sequences from DNA and protein sequence databases, and for analysing DNA and protein sequences.</p>
<p>To use function from the SeqinR package, we first need to install the SeqinR package (for instructions on how to install an R package, see <code>How to install an R package &lt;./installr.html#how-to-install-an-r-package&gt;</code>_). Once you have installed the “SeqinR” R package, you can load the “SeqinR” R package by typing:</p>
<p>.. highlight:: r</p>
<p>::</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;seqinr&quot;</span>)</code></pre></div>
<p>Remember that you can ask for more information about a particular R command by using the help() function. For example, to ask for more information about the library() function, you can type:</p>
<p>::</p>
<pre><code>&gt; help(&quot;library&quot;)</code></pre>
</div>
<div id="fasta-format" class="section level2">
<h2><span class="header-section-number">2.3</span> FASTA format</h2>
<p>The FASTA format is a simple and widely used format for storing biological (DNA or protein) sequences. It was first used by the FASTA program for sequence alignment. It begins with a single-line description starting with a “&gt;” character, followed by lines of sequences. Here is an example of a FASTA file:</p>
<p>::</p>
<pre><code>&gt; A06852 183 residues
MPRLFSYLLGVWLLLSQLPREIPGQSTNDFIKACGRELVRLWVEICGSVSWGRTALSLEE
PQLETGPPAETMPSSITKDAEILKMMLEFVPNLPQELKATLSERQPSLRELQQSASKDSN
LNFEEFKKIILNRQNEAEDKSLLELKNLGLDKHSRKKRLFRMTLSEKCCQVGCIRKDIAR
LC</code></pre>
</div>
<div id="the-ncbi-sequence-database" class="section level2">
<h2><span class="header-section-number">2.4</span> The NCBI sequence database</h2>
<p>The National Centre for Biotechnology Information (NCBI) (<code>www.ncbi.nlm.nih.gov &lt;http://www.ncbi.nlm.nih.gov/&gt;</code><em>) in the US maintains a huge database of all the DNA and protein sequence data that has been collected, the NCBI Sequence Database. This also a similar database in Europe, the European Molecular Biology Laboratory (EMBL) Sequence Database (<code>www.ebi.ac.uk/embl &lt;http://www.ebi.ac.uk/embl/&gt;</code></em>), and also a similar database in Japan, the DNA Data Bank of Japan (DDBJ; <code>www.ddbj.nig.ac.jp &lt;http://www.ddbj.nig.ac.jp/&gt;</code>_). These three databases exchange data every night, so at any one point in time, they contain almost identical data.</p>
<p>Each sequence in the NCBI Sequence Database is stored in a separate <em>record</em>, and is assigned a unique identifier that can be used to refer to that sequence record. The identifier is known as an <em>accession</em>, and consists of a mixture of numbers and letters. For example, Dengue virus causes <code>Dengue fever &lt;http://apps.who.int/tdr/svc/diseases/dengue&gt;</code>_, which is classified as a neglected tropical disease by the WHO. by any one of four types of Dengue virus: DEN-1, DEN-2, DEN-3, and DEN-4. The NCBI accessions for the DNA sequences of the DEN-1, DEN-2, DEN-3, and DEN-4 Dengue viruses are NC_001477, NC_001474, NC_001475 and NC_002640, respectively.</p>
<p>Note that because the NCBI Sequence Database, the EMBL Sequence Database, and DDBJ exchange data every night, the DEN-1 (and DEN-2, DEN-3, DEN-4) Dengue virus sequence will be present in all three databases, but it will have different accessions in each database, as they each use their own numbering systems for referring to their own sequence records.</p>
</div>
<div id="retrieving-genome-sequence-data-via-the-ncbi-website" class="section level2">
<h2><span class="header-section-number">2.5</span> Retrieving genome sequence data via the NCBI website</h2>
<p>You can easily retrieve DNA or protein sequence data from the NCBI Sequence Database via its website <code>www.ncbi.nlm.nih.gov &lt;http://www.ncbi.nlm.nih.gov/&gt;</code>_.</p>
<p>The Dengue DEN-1 DNA sequence is a viral DNA sequence, and as mentioned above, its NCBI accession is NC_001477. To retrieve the DNA sequence for the Dengue DEN-1 virus from NCBI, go to the NCBI website, type “NC_001477” in the Search box at the top of the webpage, and press the “Search” button beside the Search box:</p>
<p>|image0|</p>
<p>On the results page you will see the number of hits to “NC_001477” in each of the NCBI databases on the NCBI website. There are many databases on the NCBI website, for example, the “PubMed” data contains abstracts from scientific papers, the “Nucleotide” database contains DNA and RNA sequence data, the “Protein” data contains protein sequence data, and so on. The picture below shows what the results page should look like for your NC_001477 search. As you are looking for the DNA sequence of the Dengue DEN-1 virus<br />
genome, you expect to see a hit in the NCBI Nucleotide database, and indeed there is hit in the Nucleotide database (indicated by the “1” beside the icon for the Nucleotide database):</p>
<p>|image1|</p>
<p>To look at the one sequence found in the Nucleotide database, you need to click on the icon for the NCBI Nucleotide database on the results page for the search:</p>
<p>|image2|</p>
<p>When you click on the icon for the NCBI Nucleotide database, it will bring you to the record for NC_001477 in the NCBI Nucleotide database. This will contain the name and NCBI accession of the sequence, as well as other details such as any papers describing the sequence:</p>
<p>|image3|</p>
<p>To retrieve the DNA sequence for the DEN-1 Dengue virus genome sequence as a FASTA format sequence file, click on “Send” at the top right of the NC_001477 sequence record webpage, and then choose “File” in the pop-up menu that appears, and then choose FASTA from the “Format” menu that appears, and click on “Create file”.</p>
<p>A box will pop up asking you what to name the file, and where to save it. You should give it a sensible name (eg. “den1.fasta”) and save it in a place where you will remember (eg. in the “My Documents” folder is a good idea):</p>
<p>|image4|</p>
<p>You can now open the FASTA file containing the DEN-1 Dengue virus genome sequence using WordPad on your computer. To open WordPad, click on “Start” on the bottom left of your screen, click on “All Programs” in the menu that appears, and then select “Accessories” from the menu that appears next, and then select “WordPad” from the menu that appears next. WordPad should start up. In Wordpad, choose “Open” from the “File” menu. The WordPad “Open” dialog will appear. Set “Files of type” to “All Documents” at the bottom of the WordPad “Open” dialog. You should see a list of files, now select the file that contains the DEN-1 Dengue virus sequence (eg. “den1.fasta”). The contents of the FASTA format file containing the Dengue DEN-1 sequence should now be displayed in WordPad:</p>
<p>|image5|</p>
</div>
<div id="retrieving-genome-sequence-data-using-seqinr" class="section level2">
<h2><span class="header-section-number">2.6</span> Retrieving genome sequence data using SeqinR</h2>
<p>Instead of going to the NCBI website to retrieve sequence data from the NCBI database, you can retrieve sequence data from NCBI directly from R, by using the SeqinR R package.</p>
<p>For example, you learnt above how to retrieve the DEN-1 Dengue virus genome sequence, which has NCBI accession NC_001477, from the NCBI website. To retrieve a sequence with a particular NCBI accession, you can use R function “getncbiseq()” below, which you will first need to copy and paste into R:</p>
<p>::</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">getncbiseq &lt;-<span class="st"> </span><span class="cf">function</span>(accession)
{
   <span class="kw">require</span>(<span class="st">&quot;seqinr&quot;</span>) <span class="co"># this function requires the SeqinR R package</span>
   <span class="co"># first find which ACNUC database the accession is stored in:</span>
   dbs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;genbank&quot;</span>,<span class="st">&quot;refseq&quot;</span>,<span class="st">&quot;refseqViruses&quot;</span>,<span class="st">&quot;bacterial&quot;</span>)
   numdbs &lt;-<span class="st"> </span><span class="kw">length</span>(dbs)
   <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>numdbs)
   {
      db &lt;-<span class="st"> </span>dbs[i]
      <span class="kw">choosebank</span>(db)
      <span class="co"># check if the sequence is in ACNUC database &#39;db&#39;:</span>
      resquery &lt;-<span class="st"> </span><span class="kw">try</span>(<span class="kw">query</span>(<span class="st">&quot;.tmpquery&quot;</span>, <span class="kw">paste</span>(<span class="st">&quot;AC=&quot;</span>, accession)), <span class="dt">silent =</span> <span class="ot">TRUE</span>)
      <span class="cf">if</span> (<span class="op">!</span>(<span class="kw">inherits</span>(resquery, <span class="st">&quot;try-error&quot;</span>)))
      {
         queryname &lt;-<span class="st"> &quot;query2&quot;</span>
         thequery &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;AC=&quot;</span>,accession,<span class="dt">sep=</span><span class="st">&quot;&quot;</span>)
         <span class="kw">query</span>(<span class="st">`</span><span class="dt">queryname</span><span class="st">`</span>,<span class="st">`</span><span class="dt">thequery</span><span class="st">`</span>)
         <span class="co"># see if a sequence was retrieved:</span>
         seq &lt;-<span class="st"> </span><span class="kw">getSequence</span>(query2<span class="op">$</span>req[[<span class="dv">1</span>]])
         <span class="kw">closebank</span>()
         <span class="kw">return</span>(seq)
      }
      <span class="kw">closebank</span>()
   }
   <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;ERROR: accession&quot;</span>,accession,<span class="st">&quot;was not found&quot;</span>))
}</code></pre></div>
<p>Once you have copied and pasted the function getncbiseq() into R, you can use it to retrieve a sequence from the NCBI Nucleotide database, such as the sequence for the DEN-1 Dengue virus (accession NC_001477):</p>
<p>::</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dengueseq &lt;-<span class="st"> </span><span class="kw">getncbiseq</span>(<span class="st">&quot;NC_001477&quot;</span>)</code></pre></div>
<pre><code>## Error in getSequence(query2$req[[1]]): 객체 &#39;query2&#39;를 찾을 수 없습니다</code></pre>
<p>The variable <em>dengueseq</em> is a vector containing the nucleotide sequence. Each element of the vector contains one nucleotide of the sequence. Therefore, to print out a certain subsequence of the sequence, we just need to type the name of the vector <em>dengueseq</em> followed by the square brackets containing the indices for those nucleotides. For example, the following command prints out the first 50 nucleotides of the DEN-1 Dengue virus genome sequence:</p>
<p>::</p>
<pre><code>&gt; dengueseq[1:50]
[1] &quot;a&quot; &quot;g&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;g&quot; &quot;t&quot; &quot;c&quot; &quot;t&quot; &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot; &quot;g&quot; &quot;g&quot; &quot;a&quot;
[20] &quot;c&quot; &quot;c&quot; &quot;g&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;g&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;c&quot; &quot;g&quot; &quot;a&quot;
[39] &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;g&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;g&quot; &quot;c&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot;
 </code></pre>
<p>Note that <em>dengueseq[1:50]</em> refers to the elements of the vector <em>dengueseq</em> with indices from 1-50. These elements contain the first 50 nucleotides of the DEN-1 Dengue virus sequence.</p>
</div>
<div id="writing-sequence-data-out-as-a-fasta-file" class="section level2">
<h2><span class="header-section-number">2.7</span> Writing sequence data out as a FASTA file</h2>
<p>If you have retrieved a sequence from the NCBI database using the “getncbiseq()” function, you may want to save the sequence to a FASTA-format file on your computer, in case you need the sequence for further analyses (either in R or in other software).</p>
<p>You can write out a sequence to a FASTA-format file in R by using the “write.fasta()” function from the SeqinR R package. The write.fasta() function requires that you tell it the name of the output file using the “file.out” argument (input). You also need to specify the R variable that contains the sequence using the “sequences” argument, and the name that you want to give to the sequence using the “names” argument.</p>
<p>For example, if you have stored the DEN-1 Dengue virus sequence in a vector <em>dengueseq</em>, to write out the sequence to a FASTA-format file called “den1.fasta” that contains the sequence labelled as “DEN-1”, you can type:</p>
<p>::</p>
<pre><code>&gt; write.fasta(names=&quot;DEN-1&quot;, sequences=dengueseq, file.out=&quot;den1.fasta&quot;)</code></pre>
</div>
<div id="reading-sequence-data-into-r" class="section level2">
<h2><span class="header-section-number">2.8</span> Reading sequence data into R</h2>
<p>Using the SeqinR package in R, you can easily read a DNA sequence from a FASTA file into R. For example, we described above how to retrieve the DEN-1 Dengue virus genome sequence from the NCBI database, or from R using the getncbiseq() function, and save it in a FASTA format file (eg. “den1.fasta”).</p>
<p>You can read this FASTA format file into R using the read.fasta() function from the SeqinR R package:</p>
<p>::</p>
<pre><code>&gt; library(&quot;seqinr&quot;) 
&gt; dengue &lt;- read.fasta(file = &quot;den1.fasta&quot;)</code></pre>
<p>Note that R expects the files that you read in (eg. “den1.fasta”) to be in the “My Documents” folder on your computer, so if you stored “den1.fasta” somewhere else, you will have to move or copy it into “My Documents”.</p>
<p>The command above reads the contents of the fasta format file den1.fasta into an R object called <em>dengue</em>. The variable <em>dengue</em> is an R list object. As explained above, a list is an R object that is like a vector, but can contain elements that are numeric and/or contain characters. In this case, the list <em>dengue</em> contains information from the FASTA file that you have read in (ie.<br />
the name given to the dengue sequence in the FASTA file, and the DNA sequence itself). In fact, the first element of the list object <em>dengue</em> contains the the DNA sequence. As described above, we can access the elements of an R list object using double square brackets. Thus, we can store the DNA sequence for DEN-1 Dengue virus in a variable <em>dengueseq</em> by typing:</p>
<p>::</p>
<pre><code>&gt; dengueseq &lt;- dengue[[1]]</code></pre>
<p>Now the variable <em>dengueseq</em> is a vector containing the nucleotide sequence.</p>
</div>
<div id="length-of-a-dna-sequence" class="section level2">
<h2><span class="header-section-number">2.9</span> Length of a DNA sequence</h2>
<p>Once you have retrieved a DNA sequence, we can obtain some simple statistics to describe that sequence, such as the sequence’s total length in nucleotides. In the above example, we retrieved the DEN-1 Dengue virus genome sequence, and stored it in the vector variable <em>dengueseq</em>. To subsequently obtain the length of the genome sequence, we would use the length() function, typing:</p>
<p>::</p>
<pre><code>&gt; length(dengueseq)
[1] 10735 </code></pre>
<p>The length() function will give you back the length of the sequence stored in variable <em>dengueseq</em>, in nucleotides. The length() function actually gives the number of elements in the input vector that you pass to it, which in this case in the number of elements in the vector <em>dengueseq</em>. Since each element of the vector <em>dengueseq</em> contains one nucleotide of the DEN-1 Dengue virus<br />
sequence, the result for the DEN-1 Dengue virus genome tells us the length of its genome sequence (ie. 10735 nucleotides long).</p>
</div>
<div id="base-composition-of-a-dna-sequence" class="section level2">
<h2><span class="header-section-number">2.10</span> Base composition of a DNA sequence</h2>
<p>An obvious first analysis of any DNA sequence is to count the number of occurrences of the four different nucleotides (“A”, “C”, “G”, and “T”) in the sequence. This can be done using the the table() function. For example, to find the number of As, Cs, Gs, and Ts in the DEN-1 Dengue virus sequence (which you have put into vector variable <em>dengueseq</em>, using the commands above), you would type:</p>
<p>::</p>
<pre><code>&gt; table(dengueseq)
dengueseq
  a    c    g    t 
3426 2240 2770 2299 </code></pre>
<p>This means that the DEN-1 Dengue virus genome sequence has 3426 As, 2240 Cs, 2770 Gs and 2299 Ts.</p>
</div>
<div id="gc-content-of-dna" class="section level2">
<h2><span class="header-section-number">2.11</span> GC Content of DNA</h2>
<p>One of the most fundamental properties of a genome sequence is its GC content, the fraction of the sequence that consists of Gs and Cs, ie. the %(G+C).</p>
<p>The GC content can be calculated as the percentage of the bases in the genome that are Gs or Cs. That is, GC content = (number of Gs + number of Cs)*100/(genome length). For example, if the genome is 100 bp, and 20 bases are Gs and 21 bases are Cs, then the GC content is (20 + 21)*100/100 = 41%.</p>
<p>You can easily calculate the GC content based on the number of As, Gs, Cs, and Ts in the genome sequence. For example, for the DEN-1 Dengue virus genome sequence, we know from using the table() function above that the genome contains 3426 As, 2240 Cs, 2770 Gs and 2299 Ts. Therefore, we can calculate the GC content using the command:</p>
<p>::</p>
<pre><code>&gt; (2240+2770)*100/(3426+2240+2770+2299)
[1] 46.66977 </code></pre>
<p>Alternatively, if you are feeling lazy, you can use the GC() function in the SeqinR package, which gives the fraction of bases in the sequence that are Gs or Cs.</p>
<p>::</p>
<pre><code>&gt; GC(dengueseq)
[1] 0.4666977 </code></pre>
<p>The result above means that the fraction of bases in the DEN-1 Dengue virus genome that are Gs or Cs is 0.4666977. To convert the fraction to a percentage, we have to multiply by 100, so the GC content as a percentage is 46.66977%.</p>
</div>
<div id="dna-words" class="section level2">
<h2><span class="header-section-number">2.12</span> DNA words</h2>
<p>As well as the frequency of each of the individual nucleotides (“A”, “G”, “T”, “C”) in a DNA sequence, it is also interesting to know the frequency of longer DNA “words”. The individual nucleotides are DNA words that are 1 nucleotide long, but we may also want to find out the frequency of DNA words that are 2 nucleotides long (ie. “AA”, “AG”, “AC”, “AT”, “CA”, “CG”, “CC”, “CT”, “GA”, “GG”, “GC”, “GT”, “TA”, “TG”, “TC”, and “TT”), 3 nucleotides long (eg. “AAA”, “AAT”, “ACG”, etc.), 4 nucleotides long, etc.</p>
<p>To find the number of occurrences of DNA words of a particular length, we can use the count() function from the R SeqinR package. For example, to find the number of occurrences of DNA words that are 1 nucleotide long in the sequence <em>dengueseq</em>, we type:</p>
<p>::</p>
<pre><code>&gt; count(dengueseq, 1)
  a    c    g    t 
 3426 2240 2770 2299 </code></pre>
<p>As expected, this gives us the number of occurrences of the individual nucleotides. To find the number of occurrences of DNA words that are 2 nucleotides long, we type:</p>
<p>::</p>
<pre><code>&gt; count(dengueseq, 2)
  aa   ac   ag   at   ca   cc   cg   ct   ga   gc   gg   gt   ta   tc   tg   tt 
 1108  720  890  708  901  523  261  555  976  500  787  507  440  497  832  529 </code></pre>
<p>Note that by default the count() function includes all overlapping DNA words in a sequence. Therefore, for example, the sequence “ATG” is considered to contain two words that are two nucleotides long: “AT” and “TG”.</p>
<p>If you type help(‘count’), you will see that the result (output) of the function count() is a <em>table</em> object. This means that you can use double square brackets to extract the values of elements from the table. For example, to extract the value of the third element (the number of Gs in the DEN-1 Dengue virus sequence), you can type:</p>
<p>::</p>
<pre><code>&gt; denguetable &lt;- count(dengueseq,1)
&gt; denguetable[[3]] 
 [1] 2770</code></pre>
<p>The command above extracts the third element of the table produced by count(dengueseq,1), which we have stored in the table variable <em>denguetable</em>.</p>
<p>Alternatively, you can find the value of the element of the table in column “g” by typing:</p>
<p>::</p>
<pre><code>&gt; denguetable[[&quot;g&quot;]]
 [1] 2770</code></pre>
</div>
<div id="summary" class="section level2">
<h2><span class="header-section-number">2.13</span> Summary</h2>
<p>In this practical, you will have learnt to use the following R functions:</p>
<ol>
<li>length() for finding the length of a vector or list</li>
<li>table() for printing out a table of the number of occurrences of each type of item in a vector or list.</li>
</ol>
<p>These functions belong to the standard installation of R.</p>
<p>You have also learnt the following R functions that belong to the SeqinR package:</p>
<ol>
<li>GC() for calculating the GC content for a DNA sequence</li>
<li>count() for calculating the number of occurrences of DNA words of a particular length in a DNA sequence</li>
</ol>
</div>
<div id="links-and-further-reading" class="section level2">
<h2><span class="header-section-number">2.14</span> Links and Further Reading</h2>
<p>Some links are included here for further reading.</p>
<p>For background reading on DNA sequence statistics, it is recommended to read Chapter 1 of <em>Introduction to Computational Genomics: a case studies approach</em> by Cristianini and Hahn (Cambridge University Press; <code>www.computational-genomics.net/book/ &lt;http://www.computational-genomics.net/book/&gt;</code>_).</p>
<p>For more in-depth information and more examples on using the SeqinR package for sequence analysis, look at the SeqinR documentation, <code>http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng &lt;http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng&gt;</code>_.</p>
<p>There is also a very nice chapter on “Analyzing Sequences”, which includes examples of using SeqinR for sequence analysis, in the book <em>Applied statistics for bioinformatics using R</em> by Krijnen (available online at <code>cran.r-project.org/doc/contrib/Krijnen-IntroBioInfStatistics.pdf &lt;http://cran.r-project.org/doc/contrib/Krijnen-IntroBioInfStatistics.pdf&gt;</code>_).</p>
<p>For a more in-depth introduction to R, a good online tutorial is available on the “Kickstarting R” website, <code>cran.r-project.org/doc/contrib/Lemon-kickstart &lt;http://cran.r-project.org/doc/contrib/Lemon-kickstart/&gt;</code>_.</p>
<p>There is another nice (slightly more in-depth) tutorial to R available on the “Introduction to R” website, <code>cran.r-project.org/doc/manuals/R-intro.html &lt;http://cran.r-project.org/doc/manuals/R-intro.html&gt;</code>_.</p>
</div>
<div id="acknowledgements" class="section level2">
<h2><span class="header-section-number">2.15</span> Acknowledgements</h2>
<p>Thank you to Noel O’Boyle for helping in using Sphinx, <code>http://sphinx.pocoo.org &lt;http://sphinx.pocoo.org&gt;</code><em>, to create this document, and github, <code>https://github.com/ &lt;https://github.com/&gt;</code></em>, to store different versions of the document as I was writing it, and readthedocs, <code>http://readthedocs.org/ &lt;http://readthedocs.org/&gt;</code>_, to build and distribute this document.</p>
<p>Many of the ideas for the examples and exercises for this chapter were inspired by the Matlab case studies on <em>Haemophilus influenzae</em> (<code>www.computational-genomics.net/case\_studies/haemophilus\_demo.html &lt;http://www.computational-genomics.net/case_studies/haemophilus_demo.html&gt;</code><em>) and Bacteriophage lambda (<code>http://www.computational-genomics.net/case\_studies/lambdaphage\_demo.html &lt;http://www.computational-genomics.net/case_studies/lambdaphage_demo.html&gt;</code></em>) from the website that accompanies the book <em>Introduction to Computational Genomics: a case studies approach</em> by Cristianini and Hahn (Cambridge University Press; <code>www.computational-genomics.net/book/ &lt;http://www.computational-genomics.net/book/&gt;</code>_).</p>
<p>Thank you to Jean Lobry and Simon Penel for helpful advice on using the SeqinR package.</p>
</div>
<div id="contact" class="section level2">
<h2><span class="header-section-number">2.16</span> Contact</h2>
<p>I will be grateful if you will send me (<code>Avril Coghlan &lt;http://www.sanger.ac.uk/research/projects/parasitegenomics/&gt;</code>_) corrections or suggestions for improvements to my email address <a href="mailto:alc@sanger.ac.uk">alc@sanger.ac.uk</a></p>
</div>
<div id="license" class="section level2">
<h2><span class="header-section-number">2.17</span> License</h2>
<p>The content in this book is licensed under a <code>Creative Commons Attribution 3.0 License &lt;http://creativecommons.org/licenses/by/3.0/&gt;</code>_.</p>
</div>
<div id="exercises" class="section level2">
<h2><span class="header-section-number">2.18</span> Exercises</h2>
<p>Answer the following questions, using the R package. For each question, please record your answer, and what you typed into R to get this answer.</p>
<p>Model answers to the exercises are given in <code>Answers to the exercises on DNA Sequence Statistics (1) &lt;./chapter_answers.html#dna-sequence-statistics-1&gt;</code>_.</p>
<p>Q1. What are the last twenty nucleotides of the Dengue virus genome sequence?</p>
<p>Q2. What is the length in nucleotides of the genome sequence for the bacterium <em>Mycobacterium leprae</em> strain TN (accession NC_002677)? Note: <em>Mycobacterium leprae</em> is a bacterium that is responsible for causing <code>leprosy &lt;http://apps.who.int/tdr/svc/diseases/leprosy&gt;</code>_, which is classified by the WHO as a neglected tropical disease. As the genome sequence is a DNA sequence, if you are retrieving its sequence via the NCBI website, you will need to look for it in the NCBI Nucleotide database.</p>
<p>Q3. How many of each of the four nucleotides A, C, T and G, and any other symbols, are there in the <em>Mycobacterium leprae</em> TN genome sequence? Note: other symbols apart from the four nucleotides A/C/T/G may appear in a sequence. They correspond to positions in the sequence that are are not clearly one base or another and they are due, for example, to sequencing uncertainties. or example, the symbol ‘N’ means ‘aNy base’, while ‘R’ means ‘A or G’ (puRine). There is a table of symbols at <code>www.bioinformatics.org/sms/iupac.html &lt;http://www.bioinformatics.org/sms/iupac.html&gt;</code>_.</p>
<p>Q4. What is the GC content of the <em>Mycobacterium leprae</em> TN genome sequence, when (i) all non-A/C/T/G nucleotides are included, (ii) non-A/C/T/G nucleotides are discarded? Hint: look at the help page for the GC() function to find out how it deals with non-A/C/T/G nucleotides.</p>
<p>Q5. How many of each of the four nucleotides A, C, T and G are there in the complement of the <em>Mycobacterium leprae</em> TN genome sequence? Hint: you will first need to search for a function to calculate the complement of a sequence. Once you have found out what function to use, remember to use the help() function to find out what are the arguments (inputs) and results (outputs) of that function. How does the function deal with symbols other than the four nucleotides A, C, T and G? Are the numbers of As, Cs, Ts, and Gs in the complementary sequence what you would expect?</p>
<p>Q6. How many occurrences of the DNA words CC, CG and GC occur in the <em>Mycobacterium leprae</em> TN genome sequence?</p>
<p>Q7. How many occurrences of the DNA words CC, CG and GC occur in the (i) first 1000 and (ii) last 1000 nucleotides of the <em>Mycobacterium leprae</em> TN genome sequence? How can you check that the subsequence that you have looked at is 1000 nucleotides long?</p>
<p>.. |image0| image:: ../_static/P1_image0.png .. |image1| image:: ../_static/P1_image1.png :width: 900 .. |image2| image:: ../_static/P1_image2.png .. |image3| image:: ../_static/P1_image3.png .. |image4| image:: ../_static/P1_image4.png .. |image5| image:: ../_static/P1_image5.png</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
